Automattic/jetpack
# Get and store current product as a class variable for performance.

The "Upgrades" menu item already gets the site's current plan. So, instead of us calling the database again to get the plan for the WooCommerce menu item, we just reuse the data the "Upgrades" menu has already pulled.

<img src="https://user-images.githubusercontent.com/140841/144890889-fa08537a-8b2f-4272-9f3f-6e8de146b04f.png" width="300" />

Adds `get_current_product()` to `WPcom_Admin_Menu` class. `get_current_product()` stores the current site's plan in a class variable, so it can be referenced multiple times without hitting WPCOM_Store_API (and the database) more than once per request.

Uses our new `get_current_product()` function to pass the product data to the `jetpack_show_wpcom_woocommerce_installation_menu` filter, and inside `add_upgrades_menu()`.

Discussion and Issue: # 8-gh-Automattic/woo-on-plans

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
Ideally, you'll have Free, Personal, Premium, and Business simple test sites (with `woop` blog stickers disabled).

- Apply this PR's Fusion patch to your Sandbox. `arc patch D71072`
- Go to wp-admin on any of your simple test sites. There should be no errors (and no WooCommerce menu link, unless you have a `woop` blog sticker enabled).
- Keep this patch applied and additionally apply the corresponding diff D69877-code to the wpcom filter. `arc patch D69877`
- Sandbox Calypso.
- Point your simple test sites to your sandbox via your host file.
- You should NOT see the WooCommerce menu item on your Free, Personal, or Premium test sites, either via Calypso or wp-admin. Example URLs:
  - http://calypso.localhost:3000/home/yourtestsite.wordpress.com
  - https://yourtestsite.wordpress.com/wp-admin/
- You should see the WooCommerce menu item on Business plan sites.

Regression test:
- You should continue to see the correct plan name beside "Upgrades" in the menu.
- With this Jetpack PR applied to an Atomic site ([one method to test Atomic](pb5gDS-1rQ-p2)) `git checkout add/get-current-product-function`, you should see no log errors, and the WooCommerce installation menu item will be displayed according to the `jetpack_show_wpcom_woocommerce_installation_menu` filter criteria in WPCOMSH. Currently, that means all Atomic sites see the WooCommerce installation menu item unless WooCommerce is already installed.
